sympy/polys/matrices/normalforms.py

```diff
@@ -30,12 +37,35 @@ def smith_normal_form(m):
     ...                   [ZZ(2), ZZ(16), ZZ(14)]], (3, 3), ZZ)
     >>> print(smith_normal_form(m).to_Matrix())
     Matrix([[1, 0, 0], [0, 10, 0], [0, 0, -30]])
-
     '''
-    invs = invariant_factors(m)
-    smf = DomainMatrix.diag(invs, m.domain, m.shape)
-    return smf
-
+    '''
+    Return the Smith Normal Form of a matrix `m` along with the unimodular matrices S and T.
+    '''
+    from sympy import Matrix
+
+    M = Matrix(m)
+    S = Matrix.eye(M.rows)
+    T = Matrix.eye(M.cols)
+    J = M.copy()
+    for i in range(min(J.rows, J.cols)):
+        while J[i,i] != J[:i+1,i+1:].applyfunc(lambda x: x % J[i,i]).det():
+            non_zero_elements = [x for x in J[:,i].tolist() if x[0]]
+            g, s, t = extended_gcd(J[i,i], min(non_zero_elements))
+            s_row = [x[0] for x in J[:,i].tolist()].index(min(non_zero_elements))
+            T.row_swap(i, s_row)
+            J = J.row_swap(i, s_row)
+            J[:,i] = s*J[:,i] + t*J[:,s_row]
+        for j in range(i+1, J.rows):
+            q, r = divmod(J[j,i], J[i,i])
+            if r:
+                S.row_add(j, i, -q)
+                J.row_add(j, i, -q)
+        for j in range(i+1, J.cols):
+            q, r = divmod(J[i,j], J[i,i])
+            if r:
+                T.col_add(j, i, -q)
+                J.col_add(j, i, -q)
+    return J, S, T
```

Comment:
If we change the result to 3-tuple, it is not backward compatible and we need to provide it with different option or or different function.
I usually like to name the method like this with `_decomposition` suffix